File: server/schemas/game/action.py
from enum import Enum
from typing import Any
from pydantic import BaseModel
from schemas.game.enums import ZoneType
from schemas.game.entity import Entity
from schemas.game.class_info import CardInfo

class ActionType(Enum):
    MOVE = "move"
    SET_CARD = "set_card"
    SET_PLAYER = "set_player"
    SET_GAME = "set_game"
    ATTACK = "attack"
    DAMAGE = "damege"
    EFFECT = "effect"

class SetCardType(Enum):
    NAME = "name"
    CLASS = "class"
    ATTACK = "attack"
    HEALTH = "health"
    IMAHE_PATH = "image_path"
    TYPE = "type"
    EFEECTS = "effects"
    
class SetPlayerType(Enum):
    COST = "cost"
    ATTACK = "attack"
    HEALTH = "health"
    EFEECTS = "effects"

class SetGameType(Enum):
    EFEECTS = "effects"

class Action(BaseModel):
    action_type: ActionType
    action_data: Any

class ActionSetCard(BaseModel):
    card: Entity
    type: SetCardType
    data: Any

class ActionSetPlayer(BaseModel):
    opponent: bool
    type: SetPlayerType
    data: Any

class ActionSetGame(BaseModel):
    type: SetGameType
    data: Any


class ActionMove(BaseModel):
    before: Entity | None
    after: Entity
    destroy: bool = False
    state: CardInfo | None
    

class ActionEffect(BaseModel):
    effect_id: int
    entity: Entity
    targets: list[Entity] = []

class ActionAtteck(BaseModel):
    subject: Entity
    object: Entity

class ActionDamage(BaseModel):
    entity: Entity
    damage: int

File: client/src/types/action.ts
import { ActionType, Entity, CardInfo } from "./games"

export interface Action {
  action_type: ActionType
  action_data: any
}
export interface ActionMove {
  before: Entity
  after: Entity
}
export interface ActionCardState {
  entity: Entity
  state: CardInfo
}
export interface ActionSideEffect {
  entity: Entity
  effect: number
}
export interface ActionCost {
  cost: number
}
export interface ActionEffect {
  effect: number
  subject: Entity
  targets: Entity[]
}
export interface ActionAttack {
  subject: Entity
  object: Entity
}
export interface ActionDestroy {
  entity: Entity
}
export interface ActionDamage {
  entity: Entity
  damage: number
}